### ISSUE TITLE
Native packaging

### ISSUE THREAD

**Issue by huitseeker** [deeplearning4j/nd4j#1678]
It would be great to have native packaging (deb, rpm at least, possibly [snaps](https://www.ubuntu.com/desktop/snappy)  for libnd4j, nd4j, datavec, dl4j.

This issue is here to serve as an umbrella for organizing this package generation.

**Comment by huitseeker** [comment]
Options:
- fpm : deb, rpm packager https://github.com/jordansissel/fpm
- jdeb : deb-for-java plugin https://github.com/tcurdt/jdeb
- rpm-maven plugin: http://www.mojohaus.org/rpm-maven-plugin/ 
- jrpm : http://jrpm.sourceforge.net/

**Comment by huitseeker** [comment]
Another option for libnd4j: https://schneide.wordpress.com/2013/02/11/build-a-rpm-package-using-cmake/ https://cmake.org/Wiki/CMake:CPackPackageGenerators

**Comment by wmeddie** [comment]
So this is the Debian Java policy: https://www.debian.org/doc/packaging-manuals/java-policy/x126.html Looks like it's pretty much compatible with the Red Hat Java policy except for missing the classpath-helper tool.  It'd be great if we could have the packages also symlinked or added to the default classpath in CDH or HDP.

I felt rpm-maven-plugin was very adequate but FPM Makefiles are definitely much smaller.

Using CMake's facilities for libnd4j makes the most sense.

**Comment by huitseeker** [comment]
The RedHat policy : https://fedora-java.github.io/howto/latest/

**Comment by wmeddie** [comment]
So I added a branch that shows what we can do with sbt-native-packager.  

https://github.com/deeplearning4j/nd4j/tree/wmeddie-ospackage-1/pkg

There's a few caveats however that I'd like to discuss:

### It installs to /opt/libnd4j-xxx-java/

If we follow the debian or redhat guidelines and install to /usr/share/java, there's a chance we will overwrite other packages files unless we meticulously specify the ones that already exist as dependencies.  However the ones that are in the OS repositories can be very old and might not match what we need.  

### It relies on nd4j-xxxxx-platform

The main UX I'm aiming for (let me know if I'm wrong) is so that users can list nd4j-xxx-platform as provided and not have to include it in their uber-jars.  They just put /opt/libnd4j-native-java/lib/* in their CLASSPATH.

### Need to think about integrating with spark

I'm thinking about having a separate package that also adds these jars to the HDP or CDH classpaths.  Last time I checked HDP can install multiple versions of Hadoop so I need to figure out how best to do that.

**Comment by saudet** [comment]
> It relies on nd4j-xxxxx-platform

I must be missing something. If the whole of DL4J is in native packages, there isn't going to be anything related to DL4J in the uber JARs anyway, no?

**Comment by wmeddie** [comment]
Right.  It's a big win for containers because the application jar is much smaller.  It's the same deal as Spark.

**Comment by huitseeker** [comment]
@wmeddie One of the issues is that packages, whether `deb` or `rpm`, need to specify the dependencies in our [prerequisites](https://deeplearning4j.org/buildinglocally#prerequisites). Care should also be taken to specify the build-deps. Finally, if we want to submit that to distros, we need maintainer metadata. 

This dependency information is valuable : it lets us work with the local machine to figure out if we should depend on cuda, for example.

All of his is a bit unwieldy annoying to maintain through sbt-native. It seems it's possible to stay with maven if using anything but it.

I have a [centos7 docker image](https://hub.docker.com/r/huitseeker/dl4j-centos7-dev-env/) provisioned with our dependencies if you want to test deployment.

**Comment by huitseeker** [comment]
See https://github.com/deeplearning4j/libnd4j/pull/366
